### PR TITLE
[category-feed-persoo] added package for generating Persoo category feed

### PR DESCRIPTION
**Pull request by TomasLudvik**
| Q             | A
| ------------- | ---
|Description, the reason for the PR| First step in implementing Persoo - better search for users.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://tl-persoo-category-feed.odin.shopsys.cloud
  - https://cz.tl-persoo-category-feed.odin.shopsys.cloud

### CODE REVIEW

File: packages/category-feed-persoo/.github/workflows/run-checks-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                    php-version: '8.1'
+                    extensions: bcmath, gd, intl, pdo_pgsql, redis, pgsql, zip
+                    tools: composer
+            -   name: Install Composer dependencies
+                run: composer install --optimize-autoloader --no-interaction
+            -   name: Run parallel-lint
+                run: php vendor/bin/parallel-lint ./src ./tests
+            -   name: Run Easy Coding Standards
+                run: php vendor/bin/ecs check --verbose ./src ./tests
+            -   name: Run PHPUnit
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/.github/workflows/run-checks-tests.yaml:27]: is there a reason why we do not run phpstan here? I believe we should do that in all the packages

https://github.com/shopsys/shopsys/issues/2929

File: packages/category-feed-persoo/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Contributing
+
+Thank you for your contributions to Shopsys Category Feed Persoo package.
+Together we are making Shopsys Platform better.
+
+This repository is READ-ONLY.
+If you want to [report issues](https://github.com/shopsys/shopsys/issues/new) and/or send [pull requests](https://github.com/shopsys/shopsys/compare),
+please use the main [Shopsys repository](https://github.com/shopsys/shopsys).
+
+Please, check our [Contribution Guide](https://github.com/shopsys/shopsys/blob/master/CONTRIBUTING.md) before contributing.
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/CONTRIBUTING.md:10]: I think master is not the proper branch for linking anymore, the question is, which one is? :smile: probably `14.0`, right? Another thing that should be changed across the packages :thinking: 

https://github.com/shopsys/shopsys/issues/2928

File: packages/category-feed-persoo/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# Shopsys Category Feed Persoo
+
+[![Downloads](https://img.shields.io/packagist/dt/shopsys/category-feed-persoo.svg)](https://packagist.org/packages/shopsys/category-feed-persoo)
+
+Plugin for [Shopsys Platform](https://www.shopsys-framework.com) responsible for generating XML category feed for [Persoo](https://persoo.cz).
+
+For detailed information about the feed see [official documentation](https://support.persoo.ai).
+
+This repository is maintained by [shopsys/shopsys] monorepo, information about changes is in [monorepo CHANGELOG.md](https://github.com/shopsys/shopsys/blob/master/CHANGELOG.md).
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/README.md:9]: as mentioned above, we should not link the files in the aster branch anymore

Comment (vitek-rostislav) [inline packages/category-feed-persoo/README.md:9]: as mentioned above, we should not link the files in the master branch anymore

File: packages/category-feed-persoo/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+## Contributing
+
+Thank you for your contributions to Shopsys Category Feed Persoo package.
+Together we are making Shopsys Platform better.
+
+This repository is READ-ONLY.
+If you want to [report issues](https://github.com/shopsys/shopsys/issues/new) and/or send [pull requests](https://github.com/shopsys/shopsys/compare),
+please use the main [Shopsys repository](https://github.com/shopsys/shopsys).
+
+Please, check our [Contribution Guide](https://github.com/shopsys/shopsys/blob/master/CONTRIBUTING.md) before contributing.
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/README.md:62]:
```diff
- https://github.com/shopsys/shopsys/blob/master/CONTRIBUTING.md
+ https://github.com/shopsys/shopsys/blob/14.0/CONTRIBUTING.md
```

File: packages/category-feed-persoo/src/Model/Category/PersooCategoryRepository.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    ) {
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig $domainConfig
+     * @param int|null $lastSeekId
+     * @param int $maxResults
+     * @return \Shopsys\FrameworkBundle\Model\Category\Category[]
+     */
+    public function getProducts(
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/src/Model/Category/PersooCategoryRepository.php]: should be named `getCategories` :slightly_smiling_face:

File: packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+     * @return \Shopsys\CategoryFeed\PersooBundle\Model\FeedItem\PersooCategoryFeedItem
+     */
+    public function create(Category $category, DomainConfig $domainConfig): PersooCategoryFeedItem
+    {
+        $locale = $domainConfig->getLocale();
+        $hierarchyIds = [$category->getId()];
+        $hierarchyNames = [$category->getName($locale)];
+        $parent = $category->getParent();
+        $rootCategory = $this->categoryRepository->getRootCategory();
+        $imageUrl = $this->imageFacade->getImageUrl($domainConfig, $category) === '' ? null : $this->imageFacade->getImageUrl($domainConfig, $category);
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php]: will `getImageUrl` ever return an empty string? It seems it throws `ImageNotFoundException` when the image does not exist, or am I missing something?

File: packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+     * @param \Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig $domainConfig
+     * @return \Shopsys\CategoryFeed\PersooBundle\Model\FeedItem\PersooCategoryFeedItem
+     */
+    public function create(Category $category, DomainConfig $domainConfig): PersooCategoryFeedItem
+    {
+        $locale = $domainConfig->getLocale();
+        $hierarchyIds = [$category->getId()];
+        $hierarchyNames = [$category->getName($locale)];
+        $parent = $category->getParent();
+        $rootCategory = $this->categoryRepository->getRootCategory();
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php:39]: it seems a little bit ineffective - this query will be called over and over again for each category item, maybe we can introduce an application cache for the root category here?

File: packages/category-feed-persoo/src/Resources/config/services.yaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+        autowire: true
+        autoconfigure: true
+        public: false
+
+    Shopsys\CategoryFeed\PersooBundle\:
+        resource: '../../*'
+
+    Shopsys\CategoryFeed\PersooBundle\PersooCategoryFeed:
+        tags:
+            - { name: shopsys.product_feed }
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/src/Resources/config/services.yaml]: now I am thinking - the branch should be probably rebased on `tl-run-crons-separately`, right?

Also, as you mentioned earlier, we should consider renaming the tag to something more general (`shopsys.feed`?). Even though, TBH, the current state does not feel as weird to me as I expected. It is labeled "product feed " and it is a product category feed, which is fine by me :shrug:  :smile:

Comment (vitek-rostislav) [inline packages/category-feed-persoo/src/Resources/config/services.yaml]: now I am thinking - the branch should be probably rebased on `tl-run-crons-separately`, right?

Also, as you mentioned earlier, we should consider renaming the tag to something more general (`shopsys.feed`?). Even though, TBH, the current state does not feel as weird to me as I expected. It is labeled "product feed " and it is a product category feed, which is fine by me :shrug:  :smile:

File: packages/category-feed-persoo/src/ShopsysCategoryFeedPersooBundle.php [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function build(ContainerBuilder $container)
+    {
+        parent::build($container);
+
+        $container->addCompilerPass(
+            DoctrineOrmMappingsPass::createAnnotationMappingDriver(
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/src/ShopsysCategoryFeedPersooBundle.php:21]: Do we need it? There are no entities so far - will there ever be any?

File: packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        $this->imageFacadeMock->method('getImageUrl')
+            ->with($domain, $category)->willReturn($url);
+    }
+
+    public function testMinimalPersooCategoryFeedItemIsCreatable(): void
+    {
+        $persooCategoryFeedItem = $this->persooFeedItemFactory->create($this->defaultCategory, $this->defaultDomain);
+
+        self::assertInstanceOf(PersooCategoryFeedItem::class, $persooCategoryFeedItem);
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php]: image

The inspection is right - the `create` method has a return type hint so `$persooCategoryFeedItem` will be an instance of `PersooCategoryFeedItem` without any doubt :slightly_smiling_face:

File: packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+        self::assertEquals(1, $persooCategoryFeedItem->getId());
```
Comment (vitek-rostislav) [inline packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php]: the test would be a little bit more readable if you moved the expected values into class constants (you could then use them in the setup methods as well). I do not insist on that, I just think it is a good idea :smile:

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]
Hi, nice job there! I have added a few suggestions, check them out please :wink:

File: packages/category-feed-persoo/.github/workflows/run-checks-tests.yaml (hunk 1 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/.github/workflows/run-checks-tests.yaml:27]: Let's solve it as a whole in the mentioned issue.

File: packages/category-feed-persoo/CONTRIBUTING.md (hunk 2 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/CONTRIBUTING.md:10]: Let's solve it as a whole in the mentioned issue.

File: packages/category-feed-persoo/README.md (hunk 3 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/README.md:9]: Let's solve it as a whole in the mentioned issue.

Comment (TomasLudvik) [reply packages/category-feed-persoo/README.md:9]: Let's solve it as a whole in the mentioned issue.

File: packages/category-feed-persoo/README.md (hunk 4 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/README.md:62]: Let's solve it as a whole in the mentioned issue.

File: packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php (hunk 7 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php:39]: I have added cache for that

File: packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php (hunk 6 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/src/Model/FeedItem/PersooCategoryFeedItemFactory.php]: I have reworked it to now work correctly with throwed exception

File: packages/category-feed-persoo/src/Model/Category/PersooCategoryRepository.php (hunk 5 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/src/Model/Category/PersooCategoryRepository.php]: Renamed :)

File: packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php (hunk 10 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php]: I have removed it

File: packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php (hunk 11 above)
Comment (TomasLudvik) [reply packages/category-feed-persoo/tests/Unit/PersooCategoryFeedItemTest.php]: I have introduced several constants 👍

**Review by vitek-rostislav — APPROVED** [review]

**Pull request merged**